README.md
## spandoc

[![Build Status][actions-badge]][actions-url]
[![Latest Version][version-badge]][version-url]
[![Rust Documentation][docs-badge]][docs-url]

[actions-badge]: https://github.com/yaahc/spandoc/workflows/Continuous%20integration/badge.svg
[actions-url]: https://github.com/yaahc/spandoc/actions?query=workflow%3A%22Continuous+integration%22
[version-badge]: https://img.shields.io/crates/v/spandoc.svg
[version-url]: https://crates.io/crates/spandoc
[docs-badge]: https://img.shields.io/badge/docs-latest-blue.svg
[docs-url]: https://docs.rs/spandoc

Attribute macro that transforms doc comments in functions into tracing [`spans`](https://docs.rs/tracing/0.1.16/tracing/span/index.html).

# Details

All doc comments intended to be transformed into spans **must** begin with `SPANDOC: `:

```rust
use spandoc::spandoc;
use tracing::info;

#[spandoc]
fn foo() {
    /// SPANDOC: this will be converted into a span
    info!("event 1");

    /// this will be ignored and produce a warning for an unused doc comment
    info!("event 2");
}
```

The spans that are created by spandoc are explicitly scoped to the expression
they're associated with.

```rust
use spandoc::spandoc;
use tracing::info;

#[spandoc]
fn main() {
    tracing_subscriber::fmt::init();
    let local = 4;

    /// SPANDOC: Emit a tracing info event {?local}
    info!("event 1");

    info!("event 2");
}
```

Running the above example will produce the following output

<pre><font color="#06989A"><b>spandoc</b></font> on <font color="#75507B"><b> await-support</b></font> <font color="#CC0000"><b>[!+] </b></font>is <font color="#FF8700"><b>📦 v0.1.3</b></font> via <font color="#CC0000"><b>🦀 v1.44.1</b></font>
<font color="#4E9A06"><b>❯</b></font> cargo run --example scoped
<font color="#4E9A06"><b>    Finished</b></font> dev [unoptimized + debuginfo] target(s) in 0.03s
<font color="#4E9A06"><b>     Running</b></font> `target/debug/examples/scoped`
<font color="#A1A1A1">Jul 09 12:42:43.691 </font><font color="#4E9A06"> INFO</font> <b>main::comment{</b>local=4 text=Emit a tracing info event<b>}</b>: scoped: event 1
<font color="#A1A1A1">Jul 09 12:42:43.691 </font><font color="#4E9A06"> INFO</font> scoped: event 2</pre>

Local variables can be associated with the generated spans by adding a
trailing block to the doc comment. The syntax for fields in the span is the
[same as in `tracing`](https://docs.rs/tracing/*/tracing/index.html#using-the-macros).

```rust
use spandoc::spandoc;
use tracing::info;

#[spandoc]
fn foo() {
    let path = "fake.txt";
    /// SPANDOC: going to load config {?path}
    info!("event 1");

    /// this will be ignored and produce a warning for an unused doc comment
    info!("event 2");
}
```

When applied to expressions that contain `await`s spandoc will correctly
use `instrument()` and exit/enter the span when suspending and resuming the
future. If there are multiple await expressions inside of the annotated
expression it will instrument each expression with the same span. The macro
will not recurse into `async` blocks.


```rust
use std::future::Future;
use spandoc::spandoc;
use tracing::info;

fn make_liz() -> impl Future<Output = Result<(), ()>> {
    info!("this will be printed in the span from `clever_girl`");

    liz()
}

async fn liz() -> Result<(), ()> {
    info!("this will also be printed in the span from `clever_girl`");

    // return a result so we can call map outside of the scope of the future
    Ok(())
}

#[spandoc]
async fn clever_girl() {
    // This span will be entered before the await, exited correctly when the
    // future suspends, and instrument the future returned from `liz` with
    // `tracing-futures`
    /// SPANDOC: clever_girl async span
    make_liz().await.map(|()| info!("this will also be printed in the span"));
}
```

#### License

<sup>
Licensed under either of <a href="LICENSE-APACHE">Apache License, Version
2.0</a> or <a href="LICENSE-MIT">MIT license</a> at your option.
</sup>

<br>

<sub>
Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in this crate by you, as defined in the Apache-2.0 license, shall
be dual licensed as above, without any additional terms or conditions.
</sub>


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

<!-- next-header -->

## [Unreleased] - ReleaseDate

## [0.2.2] - 2022-04-26
### Fixed
- Update internal `FancyGuard` type to use atomics rather than `Cell` to ensure
  that futures using spandoc remain `Send`

## [0.2.1] - 2020-11-11
### Fixed
- Fixed compiler error when using `#[spandoc]` on functions containing an
  empty block

<!-- next-url -->
[Unreleased]: https://github.com/yaahc/spandoc/compare/v0.2.2...HEAD
[0.2.2]: https://github.com/yaahc/spandoc/compare/v0.2.1...v0.2.2
[0.2.1]: https://github.com/yaahc/spandoc/releases/tag/v0.2.1


tests/happy.rs
use spandoc::spandoc;
use tracing::{info, instrument};
use tracing_error::ErrorLayer;
use tracing_subscriber::{fmt::FmtLayer, layer::Layer, registry::Registry, EnvFilter};

#[spandoc]
fn spanned() {
    /// Setting a to 1
    let _a = eventful();

    /// Doing something in a block
    {
        /// Seriously, I'm gonna do it!
        let _b = eventful();
    }
}

#[instrument]
fn eventful() -> i32 {
    /// so much to do, so little time
    info!("doing the thing!");
    42
}

#[test]
fn happy() {
    let filter = EnvFilter::try_from_default_env()
        .or_else(|_| EnvFilter::try_new("info"))
        .unwrap();

    let subscriber = ErrorLayer::default()
        .and_then(FmtLayer::builder().with_target(false).finish())
        .and_then(filter)
        .with_subscriber(Registry::default());

    tracing::subscriber::set_global_default(subscriber).expect("Could not set global default");

    spanned();
}


tests/compile_tests.rs
#[test]
fn no_std() {
    let t = trybuild::TestCases::new();
    t.pass("tests/compile/with_multi_await.rs");
    t.pass("tests/compile/with_single_await.rs");
    t.pass("tests/compile/with_valid_span_args.rs");
    t.pass("tests/compile/empty_blocks.rs");
    t.compile_fail("tests/compile/with_invalid_span_args.rs");
}


tests/spandoc_is_sync.rs
#[spandoc::spandoc]
async fn test_thing() {
    /// SPANDOC: thing
    async {}.await
}

fn is_send<T: Send>(_t: T) {}

#[test]
fn test_is_send() {
    is_send(test_thing());
}


spandoc-attribute/src/fake.rs
fn blah() {
    let val = foo_fut()
        .await
        .map_err()
        .to_other_fut()
        .await
        .map(bar.await);
}

async fn wrap(&self, f: impl Future) {
    self.span.with_subscriber(|(id, sub)| sub.exit(id));
    f.instrument(self.span).await;
    self.span.with_subscriber(|(id, sub)| sub.enter(id));
}

// --- fancy_guard version courtesy of Nika ---
async fn blah() {
    let span = span!(); // create the span & stuff.
    let fancy_guard = FancyGuard::new(&span);
    let val =
        fancy_guard.wrap(
            fancy_guard.enter_after(fancy_guard.wrap(foo_fut())
                .await)
                .map_err()
                .to_other_fut()
        )
        .await
        .map(fancy_guard.wrap(bar).await);
}

// --- split on method call version ---
fn blah() {
    let val = {
        // recurse
        let fut = foo_fut().await.map_err().to_other_fut();
        // instrument
        let val = fut.await;
        // recurse
        let arg1 = bar.await;
        // enter
        val.map(arg1)
    };
}

fn blah() {
    let val = {
        let fut = {
            // enter
            let fut = foo_fut();
            // instrument
            let val = fut.await;
            // enter
            val.map_err().to_other_fut()
        };
        // instrument
        let val = fut.await;
        // recurse
        let arg1 = bar.await;
        // enter
        val.map(arg1)
    };
}

fn blah() {
    let val = {
        let fut = {
            // enter
            let fut = foo_fut();
            // instrument
            let val = fut.await;
            // enter
            val.map_err().to_other_fut()
        };
        // instrument
        let val = fut.await;
        // instrument
        let arg1 = bar.await;
        // enter
        val.map(arg1)
    };
}

fn blah_equivalent() {
    let val = {
        let fut = foo_fut();
        let val = fut.await;
        let fut = val.map_err().to_other_fut();
        let val = fut.await;
        val.map()
    };
}


src/lib.rs
//! Attribute macro that transforms doc comments in functions into tracing
//! [`spans`](https://docs.rs/tracing/0.1.16/tracing/span/index.html).
//!
//! Spandoc provides a function level attribute macro that converts doc !
//! comments _within_ a function into tracing spans. These spans are entered
//! before the annotated expression and exited immediately after. Spandoc is also
//! async aware, so expressions that contain `.await`s will correctly instrument
//! the future, exit before yielding, and enter the span again when the future
//! resumes.
//!
//! # Details
//!
//! All doc comments intended to be transformed into spans **must** begin with `SPANDOC: `:
//!
//! ```rust
//! use spandoc::spandoc;
//! use tracing::info;
//!
//! #[spandoc]
//! fn foo() {
//!     /// SPANDOC: this will be converted into a span
//!     info!("event 1");
//!
//!     /// this will be ignored and produce a warning for an unused doc comment
//!     info!("event 2");
//! }
//! ```
//!
//! The spans that are created by spandoc are explicitly scoped to the expression
//! they're associated with.
//!
//! ```rust
//! use spandoc::spandoc;
//! use tracing::info;
//!
//! #[spandoc]
//! fn main() {
//!     tracing_subscriber::fmt::init();
//!     let local = 4;
//!
//!     /// SPANDOC: Emit a tracing info event {?local}
//!     info!("event 1");
//!
//!     info!("event 2");
//! }
//! ```
//!
//! Running the above example will produce the following output
//!
//! <pre><font color="#06989A"><b>spandoc</b></font> on <font color="#75507B"><b> await-support</b></font> <font color="#CC0000"><b>[!+] </b></font>is <font color="#FF8700"><b>📦 v0.1.3</b></font> via <font color="#CC0000"><b>🦀 v1.44.1</b></font>
//! <font color="#4E9A06"><b>❯</b></font> cargo run --example scoped
//! <font color="#4E9A06"><b>    Finished</b></font> dev [unoptimized + debuginfo] target(s) in 0.03s
//! <font color="#4E9A06"><b>     Running</b></font> `target/debug/examples/scoped`
//! <font color="#A1A1A1">Jul 09 12:42:43.691 </font><font color="#4E9A06"> INFO</font> <b>main::comment{</b>local=4 text=Emit a tracing info event<b>}</b>: scoped: event 1
//! <font color="#A1A1A1">Jul 09 12:42:43.691 </font><font color="#4E9A06"> INFO</font> scoped: event 2</pre>
//!
//! Local variables can be associated with the generated spans by adding a
//! trailing block to the doc comment. The syntax for fields in the span is the
//! [same as in `tracing`](https://docs.rs/tracing/*/tracing/index.html#using-the-macros).
//!
//! ```rust
//! use spandoc::spandoc;
//! use tracing::info;
//!
//! #[spandoc]
//! fn foo() {
//!     let path = "fake.txt";
//!     /// SPANDOC: going to load config {?path}
//!     info!("event 1");
//!
//!     /// this will be ignored and produce a warning for an unused doc comment
//!     info!("event 2");
//! }
//! ```
//!
//! When applied to expressions that contain `await`s spandoc will correctly
//! use `instrument()` and exit/enter the span when suspending and resuming the
//! future. If there are multiple await expressions inside of the annotated
//! expression it will instrument each expression with the same span. The macro
//! will not recurse into `async` blocks.
//!
//!
//! ```rust
//! use std::future::Future;
//! use spandoc::spandoc;
//! use tracing::info;
//!
//! fn make_liz() -> impl Future<Output = Result<(), ()>> {
//!     info!("this will be printed in the span from `clever_girl`");
//!
//!     liz()
//! }
//!
//! async fn liz() -> Result<(), ()> {
//!     # let span = tracing::Span::current();
//!     # let metadata = span.metadata().expect("span should have metadata");
//!     # assert_eq!(metadata.name(), "clever_girl::comment");
//!     #
//!     info!("this will also be printed in the span from `clever_girl`");
//!
//!     // return a result so we can call map outside of the scope of the future
//!     Ok(())
//! }
//!
//! #[spandoc]
//! async fn clever_girl() {
//!     // This span will be entered before the await, exited correctly when the
//!     // future suspends, and instrument the future returned from `liz` with
//!     // `tracing-futures`
//!     /// SPANDOC: clever_girl async span
//!     make_liz().await.map(|()| info!("this will also be printed in the span"));
//! }
//! #
//! # tracing_subscriber::fmt::init();
//! # futures::executor::block_on(clever_girl());
//! ```
#![doc(html_root_url = "https://docs.rs/spandoc/0.2.2")]
#![cfg_attr(docsrs, feature(doc_cfg))]
#![warn(
    missing_docs,
    rustdoc::missing_doc_code_examples,
    rust_2018_idioms,
    unreachable_pub,
    bad_style,
    const_err,
    dead_code,
    improper_ctypes,
    non_shorthand_field_patterns,
    no_mangle_generic_items,
    overflowing_literals,
    path_statements,
    patterns_in_fns_without_body,
    private_in_public,
    unconditional_recursion,
    unused,
    unused_allocation,
    unused_comparisons,
    unused_parens,
    while_true
)]
#![allow(clippy::needless_doctest_main)]
use std::sync::atomic::{AtomicBool, Ordering};

use tracing_futures::Instrument;

/// Attribute macro that transforms doc comments in functions into tracing [`spans`](https://docs.rs/tracing/0.1.16/tracing/span/index.html).
pub use spandoc_attribute::spandoc;

#[doc(hidden)]
pub struct FancyGuard<'a> {
    span: &'a tracing::Span,
    entered: AtomicBool,
}

impl<'a> FancyGuard<'a> {
    #[doc(hidden)]
    pub fn new(span: &'a tracing::Span) -> FancyGuard<'a> {
        span.with_subscriber(|(id, sub)| sub.enter(id));
        Self {
            span,
            entered: AtomicBool::new(true),
        }
    }

    #[doc(hidden)]
    pub async fn wrap<F>(&self, fut: F) -> F::Output
    where
        F: std::future::Future,
    {
        self.span.with_subscriber(|(id, sub)| sub.exit(id));
        self.entered.store(false, Ordering::Relaxed);
        let output = fut.instrument(self.span.clone()).await;
        self.span.with_subscriber(|(id, sub)| sub.enter(id));
        self.entered.store(true, Ordering::Relaxed);
        output
    }
}

impl Drop for FancyGuard<'_> {
    fn drop(&mut self) {
        if self.entered.load(Ordering::Relaxed) {
            self.span.with_subscriber(|(id, sub)| sub.exit(id));
        }
    }
}
